[CUS-114] fix: Update NeptuneScaleLogger to initialize run instance only on rank 0 to prevent multiprocessing conflicts when using ddp-spawn (#4)
neptune-ai/pytorch-lightning: README.md

> # Neptune Scale Logger for PyTorch Lightning

Reviewer: ```suggestion
# Neptune Logger for PyTorch Lightning
```

[![PyTorch Lightning](https://img.shields.io/badge/PyTorch%20Lightning-2.5.1rc2-blue)](https://lightning.ai/)

> [![Neptune Scale](https://img.shields.io/badge/Neptune%20Scale-%3E%3D0.18.0-orange)](https://neptune.ai/)

Reviewer: Shouldn't this rather point to https://github.com/neptune-ai/neptune-client-scale?

> This integration allows you to log your PyTorch Lightning training runs to Neptune Scale, providing seamless experiment tracking.

Reviewer: ```suggestion
This integration allows you to log your PyTorch Lightning training runs to Neptune 3.x, providing seamless experiment tracking.
```

## Changelog

- **2025-08-11** - Updated to support `ddp-spawn` strategy
- **2025-07-25** - Updated to better support nested mappings and dataclasses in `log_hyperparameters()`
- **2025-06-27** - Initial release

## Installation

> > [!WARNING] Prerequisites

Reviewer: I don't think custom titles can be set like this

```suggestion
> [!IMPORTANT]
```

Author: ah.. it worked in the Cursor README preview :/

> This integration works only with `neptune-scale>=0.18.0`

> There are two ways this integration can be used before it is included in the official PyTorch Lightning release:

Reviewer: ```suggestion
To install the integration, use one of the following options:
```

### Option 1: Install from source (recommended)

Install the custom version of `pytorch-lightning` from the source:

```bash
pip install git+https://github.com/neptune-ai/pytorch-lightning.git
```

### Option 2: Manual file replacement

> In your `pytorch-lightning/src/lightning/pytorch/loggers` folder, replace the `__init__.py` and `neptune.py` files with the corresponding files available [here](https://github.com/SiddhantSadangi/pytorch-lightning/tree/master/src/lightning/pytorch/loggers)

Reviewer: Shouldn't this link also point to this fork?

> [!NOTE]

> > The integration is built on top of `pytorch-lightning==2.5.1rc2`

Reviewer: I'd just move this version info in with the previous callout.

```suggestion
> Once the integration is included in the official PyTorch Lightning release, it's not necessary install it separately.
```

## Quickstart

```python
from lightning.pytorch.loggers.neptune import NeptuneScaleLogger
from lightning.pytorch import Trainer

# Initialize the logger
neptune_logger = NeptuneScaleLogger()

# Use with your trainer
trainer = Trainer(
    ...,
    logger=neptune_logger,
)

> # Your training will automatically log to Neptune Scale

Reviewer: ```suggestion
# Your training will automatically log to Neptune
```

trainer.fit(model, datamodule, ...)
```

## Usage

### 1. Import the logger

```python
from lightning.pytorch.loggers.neptune import NeptuneScaleLogger
```

### 2. Initialize NeptuneScaleLogger

```python
neptune_scale_logger = NeptuneScaleLogger()
```

> [!TIP]
> - Check the class signature and docstring for initialization args.
> - Check the `Run()` init params for additional `kwargs` that can be passed to `NeptuneScaleLogger()`: [Run | neptune.ai docs](https://docs.neptune.ai/run/#parameters)

### 3. Pass to Trainer

```python
trainer = Trainer(
    logger=neptune_scale_logger,
    # ... other trainer arguments
)
```

### 4. Log metadata to Neptune

With the Neptune integration, you can combine the default PyTorch Lightning logging methods with the Neptune methods and capabilities.

#### Use Lightning's logging methods

> Use `self.log()` in your Lightning module as usual - everything will be automatically logged to Neptune Scale.

Reviewer: ```suggestion
Use `self.log()` in your Lightning module as usual – everything will be automatically logged to Neptune.
```

#### Use Neptune's logging methods
In addition to the default logging methods, you can use the Neptune [Run](https://docs.neptune.ai/run) methods to log more metadata types. You can also use the Neptune methods outside of the training loop.

To access the Neptune run directly, use the `neptune_scale_logger.run` reference:

```python
neptune_scale_logger.run.assign_files({"dataset/data_sample": "sample_data.csv"})
```

You can also use the `neptune_scale_logger.run` reference to log outside the prefix passed to `NeptuneScaleLogger`.

#### Add tags

> Use `tags` to identify and organize your runs. Add tags with the `neptune_scale_logger.run` reference:

Reviewer: ```suggestion
Use tags to identify and organize your runs. Add tags with the `neptune_scale_logger.run` reference:
```

```python
neptune_scale_logger.run.add_tags(["notebook", "lightning"])
```

#### Log hyperparameters

> To log hyperparameters, use the standard log_hyperparams() method from the PyTorch Lightning logger:

Reviewer: ```suggestion
To log hyperparameters, use the standard `log_hyperparams()` method from the PyTorch Lightning logger:
```
